swarm/src/lib.rs

```diff
@@ -1343,14 +1363,146 @@ impl<'a> PollParameters for SwarmPollParameters<'a> {
     }
 }
 
+pub struct SwarmConfig {
+    pool_config: PoolConfig,
+}
+
+impl SwarmConfig {
+    /// Creates a new [`SwarmConfig`] from the given executor. The [`Swarm`] is obtained via
+    /// [`Swarm::new_with_config`].
+    pub fn with_executor(executor: impl Executor + Send + 'static) -> Self {
+        Self {
+            pool_config: PoolConfig::new(Some(Box::new(executor))),
+        }
+    }
+
+    /// Sets executor to the `wasm` executor.
+    /// Background tasks will be executed by the browser on the next micro-tick.
+    ///
+    /// Spawning a task is similar too:
+    /// ```typescript
+    /// function spawn(task: () => Promise<void>) {
+    ///     task()
+    /// }
+    /// ```
+    #[cfg(feature = "wasm-bindgen")]
+    pub fn with_wasm_executor() -> Self {
+        Self::with_executor(crate::executor::WasmBindgenExecutor)
+    }
+
+    /// Builds a new [`SwarmConfig`] from the given `tokio` executor.
+    #[cfg(all(
+        feature = "tokio",
+        not(any(target_os = "emscripten", target_os = "wasi", target_os = "unknown"))
+    ))]
+    pub fn with_tokio_executor() -> Self {
+        Self::with_executor(crate::executor::TokioExecutor)
+    }
+
+    /// Builds a new [`SwarmConfig`] from the given `async-std` executor.
+    #[cfg(all(
+        feature = "async-std",
+        not(any(target_os = "emscripten", target_os = "wasi", target_os = "unknown"))
+    ))]
+    pub fn with_async_std_executor() -> Self {
+        Self::with_executor(crate::executor::AsyncStdExecutor)
+    }
+
+    // TODO: Should we remove this from `SwarmConfig`?!
+    //
+    /// Creates a new [`SwarmConfig`].
+    ///
+    /// ## ⚠️  Performance warning
+    /// All connections will be polled on the current task, thus quite bad performance
+    /// characteristics should be expected. Whenever possible use an executor and
+    /// [`SwarmConfig::with_executor`].
+    pub fn without_executor() -> Self {
+        Self {
+            pool_config: PoolConfig::new(None),
+        }
+    }
```

Comment:
I find this to be a valid question. Should we carry the legacy `without_executor` over to the new `SwarmConfig`? In case there is demand, we can still introduce it in the future.

Reply:
4853caf88